### PR TITLE
Fix Timestamp.round errors

### LINKED ISSUE #22591: Inconsistent behaviour in Timestamp.round
#### Code Sample

```python
>>> import pandas as pd
>>> a = pd.to_datetime('2000-01-01 0:0:0.124999360')
>>> b = pd.to_datetime('2018-01-01 0:0:0.124999360')
>>> a.round('us')
Timestamp('2000-01-01 00:00:00.124999')
>>> b.round('us')
Timestamp('2018-01-01 00:00:00.125000')
```
#### Problem description

According to the standard definition of [rounding to a multiple](https://en.wikipedia.org/wiki/Rounding#Rounding_to_a_specified_multiple) `b.round('us')`  in the example above gives the wrong answer.

I suspect that there is loss of precision due to a floating point conversion.

```
>>> round(a.value / 1e9, 6)
946684800.124999
>>> round(b.value / 1e9, 6)
1514764800.125
```

#### Expected Output

```
>>> b.round('us')
Timestamp('2018-01-01 00:00:00.124999')
```

#### Output of ``pd.show_versions()``

<details>

INSTALLED VERSIONS
------------------
commit: None
python: 3.7.0.final.0
python-bits: 64
OS: Darwin
OS-release: 17.7.0
machine: x86_64
processor: i386
byteorder: little
LC_ALL: None
LANG: en_US.UTF-8
LOCALE: en_US.UTF-8

pandas: 0.23.4
pytest: None
pip: 18.0
setuptools: 39.0.1
Cython: None
numpy: 1.15.1
scipy: None
pyarrow: None
xarray: None
IPython: None
sphinx: None
patsy: None
dateutil: 2.7.3
pytz: 2018.5
blosc: None
bottleneck: None
tables: None
numexpr: None
feather: None
matplotlib: None
openpyxl: None
xlrd: None
xlwt: None
xlsxwriter: None
lxml: None
bs4: None
html5lib: None
sqlalchemy: None
pymysql: None
psycopg2: None
jinja2: None
s3fs: None
fastparquet: None
pandas_gbq: None
pandas_datareader: None
</details>

### PR DESCRIPTION
**Pull request by miccoli**
- [x] closes #22591
- [x] tests added / passed
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry
- [x] rebase for the new `ambiguous` kwarg, resolve conflicts.

### CODE REVIEW

File: pandas/_libs/tslibs/timestamps.pyx
Comment (jbrockmendel) [inline pandas/_libs/tslibs/timestamps.pyx]: I don’t think cython recognizes RoundTwo as a type unless it is a cdef class.  Not entirely sure on that.

**Review by mroeschke — CHANGES_REQUESTED** [review]
Looks fairly good:

1) Nit: Could you rename the single variable names to something a little more descriptive (e.g. `d` to `divisor`, etc)
2) Could you add similar `DatetimeIndex` tests as well?

**Review by jreback — CHANGES_REQUESTED** [review]
can you add a whatsnew note for 0.24.0

File: pandas/tests/scalar/timestamp/test_unary_ops.py
Comment (jreback) [inline pandas/tests/scalar/timestamp/test_unary_ops.py]: are there round errors in timedelta ops as well?

File: pandas/_libs/tslibs/timestamps.pyx
Comment (miccoli) [reply pandas/_libs/tslibs/timestamps.pyx]: Yes, I meant this more as a comment than an actual directive. Fixed in 44902695f4adba3406fb4750e819a5776683cae7

File: pandas/_libs/tslibs/timestamps.pyx
Comment (miccoli) [reply pandas/_libs/tslibs/timestamps.pyx]: As you remarked above, Cython typing does not work on native enum classes. Here I prefer to maintain `enum` semantics and have *enumeration members* instances of the `RoundTo` class. Therefore I dropped the `mode: RoundTo` annotation and kept the `isinstance` check. See 44902695f4adba3406fb4750e819a5776683cae7

File: pandas/_libs/tslibs/timestamps.pyx
Comment (miccoli) [reply pandas/_libs/tslibs/timestamps.pyx]: This code is unreachable and the `NotImplementedError` was cruft from a previous implementation.

In 44902695f4adba3406fb4750e819a5776683cae7 I inserted an `assert False` to avoid the code returning `None` if a bug is introduced by changing the `RoundTo` class.

**Comment by miccoli** [comment]
- @mroeschke : I think I addressed your requests 
  1. Nit: 44902695f4adba3406fb4750e819a5776683cae7
  2. `DatetimeIndex` tests: 52329484df7ed099ee56cef86822ced6cc7ae7c3
- @jreback : sorry I had the above two commits ready in my pipeline, but I had not yet read your recent comments. I will address them later.

File: pandas/_libs/tslibs/timestamps.pyx
Comment (miccoli) [reply pandas/_libs/tslibs/timestamps.pyx]: (I'm not sure I fully understand your comment, please correct me if I'm mistaken).

Here I'm trying to avoid recursive calls still keeping DRY. 

<https://github.com/pandas-dev/pandas/blob/44902695f4adba3406fb4750e819a5776683cae7/pandas/_libs/tslibs/timestamps.pyx#L81-L91> 

The *simple* rounding modes (`RoundTo.MINUS_INFTY`, `RoundTo.PLUS_INFTY`, `RoundTo.NEAREST_HALF_PLUS_INFTY`, `RoundTo.NEAREST_HALF_MINUS_INFTY`) which can be implemented as one-liners, are inlined for readability and efficiency.

`RoundTo.NEAREST_HALF_EVEN` is more complex to implement and is not inlined:
<https://github.com/pandas-dev/pandas/blob/44902695f4adba3406fb4750e819a5776683cae7/pandas/_libs/tslibs/timestamps.pyx#L122-L132>

Some rounding modes of `round_nsint64` are never called, but I prefer to have a more general implementation here, in order to allow future extensions.

File: pandas/_libs/tslibs/timestamps.pyx
Comment (miccoli) [reply pandas/_libs/tslibs/timestamps.pyx]: The wikipedia article on [rounding](https://en.wikipedia.org/wiki/Rounding) is a handy reference.

- [`RoundTo.MINUS_INFTY`](https://en.wikipedia.org/wiki/Rounding#Rounding_down)
- [`RoundTo.PLUS_INFTY`](https://en.wikipedia.org/wiki/Rounding#Rounding_up)
- [`RoundTo.NEAREST_HALF_EVEN`](https://en.wikipedia.org/wiki/Rounding#Round_half_to_even)
- [`RoundTo.NEAREST_HALF_PLUS_INFTY`](https://en.wikipedia.org/wiki/Rounding#Round_half_up)
- [`RoundTo.NEAREST_HALF_MINUS_INFTY`](https://en.wikipedia.org/wiki/Rounding#Round_half_down)

Should I add the to the source code?

File: pandas/tests/scalar/timestamp/test_unary_ops.py
Comment (miccoli) [reply pandas/tests/scalar/timestamp/test_unary_ops.py]: Yes, also `timedeltas` are affected by rounding errors, but it is very unlikely that these will be found in real world applications.

BTW I found #22591 processing real scientific data, in which I had timestamps for 2018 measurements with about 1 µs resolution. On the contrary I cannot imagine of an application in which you have 50 years time deltas with such a small resolution.

Of course you could create a round-trip example:
```
Timedate -> Timedelta -> round() -> Timedate 
```
which gives different results with respect to 
```
Timedate -> round().
```
Nevertheless I would suggest to postpone the resolution of the timedelta issue after this PR is merged.

(Please note also that the current complex machinery, which tries to improve rounding errors, is implemented only for Timedate and not for Timedelta.)

**Comment by miccoli** [comment]
Remainder for myself:
- [ ] add whatsnew note for 0.24.0
- [ ] formatting changes requested by @jreback 
- [ ] rebase for the new `ambiguous` kwarg, resolve conflicts.

File: pandas/_libs/tslibs/timestamps.pyx
Comment (jreback) [reply pandas/_libs/tslibs/timestamps.pyx]: yes pls

File: pandas/_libs/tslibs/timestamps.pyx
Comment (jreback) [reply pandas/_libs/tslibs/timestamps.pyx]: ok, just trying to make it much simpler for a new reader here. inline what you can then.

File: pandas/tests/scalar/timestamp/test_unary_ops.py
Comment (jreback) [reply pandas/tests/scalar/timestamp/test_unary_ops.py]: no i get it, that's why I asked. Ok to fix later, esp if its not very common.

File: pandas/_libs/tslibs/timestamps.pyx
Comment (miccoli) [reply pandas/_libs/tslibs/timestamps.pyx]: doc string with refs added

**Comment by miccoli** [comment]
rebase done

File: pandas/tests/indexes/datetimes/test_scalar_compat.py
Comment (jreback) [inline pandas/tests/indexes/datetimes/test_scalar_compat.py]: can you use .format instead

**Review by jreback — APPROVED** [review]
looks good. @mroeschke can comment as well.

**Review by mroeschke — APPROVED** [review]
LGTM. Agreed that `.format` is preferable over `%` for the string error formatting.

**Comment by miccoli** [comment]
the **ci/circleci: py27_compat** failure above is a false negative, due to a networking problem during the build phase.

File: doc/source/whatsnew/v0.24.0.txt
Comment (jreback) [inline doc/source/whatsnew/v0.24.0.txt]: raise -> rise

File: pandas/_libs/tslibs/timestamps.pyx
Comment (jreback) [inline pandas/_libs/tslibs/timestamps.pyx]: you could say undefined mode

**Review by jreback — APPROVED** [review]
tiny typo. ping on green.

File: doc/source/whatsnew/v0.24.0.txt
Comment (miccoli) [reply doc/source/whatsnew/v0.24.0.txt]: thanks for catching this typo

**Comment by miccoli** [comment]
hope everything is ready for merge.

**Comment by jreback** [comment]
lgtm. waiting till CI is un-borked to merge.

**Comment by jreback** [comment]
thanks @miccoli

**Comment by miccoli** [comment]
thanks for merging

**Pull request merged**
